src/DatePicker/index.jsx

```diff
@@ -7,70 +7,379 @@
 
 const DatePicker = props =>
 {
+    const [ timestamp, setTimestamp ] = useTimestamp(
+        props.defaultValue,
+        props.value,
+    );
+    const [ gridStartState, setGridStartState ] = useState( null );
+    const [ hourValue, setHourValue ] = useState( undefined );
+    const [ minuteValue, setMinuteValue ] = useState( undefined );
+
     const cssMap = useTheme( componentName, props );
 
+
+    const gridStartTimestamp = gridStartState || $m( timestamp )
+        .startOf( props.type === 'month' ? 'year' : 'month' )
+        .valueOf();
+
+    const isUnitSelectable = (
+        itemTimestamp,
+        unit,
+    ) =>
+    {
+        const { max } = props;
+        const min = props.min || now();
+
+        if ( itemTimestamp > max ) return false;
+
+        return $m( itemTimestamp ).add( 1, unit ) > min;
+    };
+
+
+    const dayMatrix = () =>
+    {
+        const startMonth = gridStartTimestamp;
+
+        if ( !startMonth ) return;
+
+        const offset = ( $m( startMonth ).weekday() + 6 ) % 7;
+        const daysInMonth = $m( startMonth ).daysInMonth();
+
+        const days = _.range( -offset, daysInMonth ).map( dayIndex =>
+        {
+            const hasDate = dayIndex >= 0 && dayIndex < daysInMonth;
+            const label = hasDate ? String( dayIndex + 1 ) : '';
+            const value = hasDate ?
+                $m( startMonth ).add( dayIndex, 'day' ).valueOf() : null;
+
+            const isDisabled = hasDate && !isUnitSelectable(
+                value,
+                'day',
+            );
+
+            const isCurrent = hasDate &&
+                isTimestampEqual( value, now(), 'day' );
+            const isSelected = hasDate && _.isNumber( timestamp ) &&
+                isTimestampEqual( timestamp, value, 'day' );
+            return {
+                label,
+                value,
+                isCurrent,
+                isDisabled,
+                isSelected,
+            };
+        } );
+
+        return _.chunk( days, 7 );
+    };
+
+    const monthMatrix = () =>
+    {
+        const startYear = gridStartTimestamp;
+
+        if ( !startYear ) return;
+
+        const months = _.range( 0, 12 ).map( month =>
+        {
+            const label = copy.shortMonths[ month ];
+            const value = $m( startYear ).add( month, 'month' ).valueOf();
+
+            const isDisabled = !isUnitSelectable( value, 'month' );
+
+            const isCurrent = isTimestampEqual( value, now(), 'month' );
+            const isSelected = _.isNumber( timestamp ) &&
+                isTimestampEqual( timestamp, value, 'month' );
+            return {
+                label,
+                value,
+                isCurrent,
+                isDisabled,
+                isSelected,
+            };
+        } );
+
+        return _.chunk( months, 4 );
+    };
+
+
+    const monthLabel = copy.months[ $m( gridStartTimestamp ).month() ];
+
+    const yearLabel = () => $m( gridStartTimestamp ).year().toString();
```

Comment:
This should just be: 
```
const yearLabel = $m( gridStartTimestamp ).year().toString();
```

(I don’t see any reason to wrap it in a function)

Reply:
👍